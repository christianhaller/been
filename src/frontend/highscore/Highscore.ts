import type { StatWithDate } from "../../backend/interace";

class Highscore {
  private el: HTMLFormElement;
  private doc: HTMLDocument;
  constructor(document: HTMLDocument) {
    this.doc = document;
  }

  public static timeSince(currentDate: Date, date: Date): string {
    const seconds = Math.floor((currentDate.getTime() - date.getTime()) / 1000);

    const plural = (interval: number): string => {
      return `${Math.floor(interval) > 1 ? "s" : ""}`;
    };
    let interval = seconds / 31536000;

    if (interval > 1) {
      return `${Math.floor(interval)} year${plural(interval)}`;
    }
    interval = seconds / 2592000;
    if (interval > 1) {
      return `${Math.floor(interval)} month${plural(interval)}`;
    }
    interval = seconds / 86400;
    if (interval > 1) {
      return `${Math.floor(interval)} day${plural(interval)}`;
    }
    interval = seconds / 3600;
    if (interval > 1) {
      return `${Math.floor(interval)} hour${plural(interval)}`;
    }
    interval = seconds / 60;
    if (interval > 1) {
      return `${Math.floor(interval)} minute${plural(interval)}`;
    }
    return `${Math.floor(seconds)} second${plural(seconds)}`;
  }
  async init() {
    this.el = this.doc.querySelector(".highscore");
    const highscore = (await (
      await fetch("/api/highscore")
    ).json()) as StatWithDate[];
    const currentDate = new Date();

    const tableBody = this.el.querySelector("table tbody") as HTMLTableElement;

    const new_tbody = document.createElement("tbody");
    const classes = ["lg:px-4", "border", "px-1", "py-2 text-right"].join(" ");
    const emoji = ["🏆", "🥈", "🥉"];

    highscore.forEach(({ username, cities, countries, date, url }, index) => {
      const html = `
      <td class="${classes}">${
        emoji[index] ? emoji[index] : `${index + 1}.`
      }</td>
      <td class="${classes}"><a rel="noopener noreferrer nofollow" target="_blank" href="${url}">${username}</a></td>
      <td class="${classes}">${countries}</td>
      <td class="${classes}">${cities}</td>
      <td class="${classes}">${Highscore.timeSince(
        currentDate,
        new Date(date)
      )} ago</td>`;

      const row = new_tbody.insertRow();
      row.innerHTML = html;
      let className = "bg-gray-100";
      if (index % 2) {
        className = "bg-white";
      }
      row.classList.add(className);
    });
    tableBody.parentNode.replaceChild(new_tbody, tableBody);
    this.el.classList.remove("hidden");
  }
}

export { Highscore };
