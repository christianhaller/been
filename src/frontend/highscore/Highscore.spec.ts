import { Highscore } from "./Highscore";

const mockResponse = [
  {
    username: "christianhaller",
    date: 123,
    cities: 2,
    countries: 4,
    url: "url",
  },
  {
    username: "anneka",
    date: 123,
    cities: 2,
    countries: 99,
    url: "url",
  },
  {
    username: "another user",
    date: 123,
    cities: 2,
    countries: 333,
    url: "url",
  },
  {
    username: "user4",
    date: 123,
    cities: 2,
    countries: 100,
    url: "url",
  },
];

describe("Highscore", () => {
  it("renders table", async () => {
    const mockJsonPromise = Promise.resolve(mockResponse);
    const mockFetchPromise = Promise.resolve({
      ok: true,
      json: () => mockJsonPromise,
    });
    global.fetch = jest.fn().mockImplementation(() => mockFetchPromise);

    await new Highscore(document).init();
    expect(document.querySelector(".highscore").innerHTML).toMatchSnapshot();
  });

  it("formats", async () => {
    let sut = Highscore.timeSince(
      new Date(1604052601425),
      new Date(1604052601425 - 3600 * 1000)
    );
    expect(sut).toMatchSnapshot();

    sut = Highscore.timeSince(
      new Date(1604052601425),
      new Date(1604052601425 - 60 * 1000)
    );
    expect(sut).toMatchSnapshot();

    sut = Highscore.timeSince(
      new Date(1604052601425),
      new Date(1604052601425 - 1 * 1000)
    );
    expect(sut).toMatchSnapshot();

    sut = Highscore.timeSince(
      new Date(1604052601425),
      new Date(1604052601425 - 3600 * 1000 * 24)
    );
    expect(sut).toMatchSnapshot();

    sut = Highscore.timeSince(
      new Date(1604052601425),
      new Date(1604052601425 - 3600 * 1000 * 25)
    );
    expect(sut).toMatchSnapshot();

    sut = Highscore.timeSince(
      new Date(1604052601425),
      new Date(1604052601425 - 3600 * 1000 * 24 * 365)
    );
    expect(sut).toMatchSnapshot();

    sut = Highscore.timeSince(
      new Date(1604052601425),
      new Date(1604052601425 - 3600 * 1000 * 24 * 366)
    );
    expect(sut).toMatchSnapshot();
  });
});
