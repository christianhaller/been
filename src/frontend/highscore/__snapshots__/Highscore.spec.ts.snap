// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Highscore formats 1`] = `"60 minutes"`;

exports[`Highscore formats 2`] = `"60 seconds"`;

exports[`Highscore formats 3`] = `"1 second"`;

exports[`Highscore formats 4`] = `"24 hours"`;

exports[`Highscore formats 5`] = `"1 day"`;

exports[`Highscore formats 6`] = `"12 months"`;

exports[`Highscore formats 7`] = `"1 year"`;

exports[`Highscore renders table 1`] = `
"<div><div class=\\"highscore\\"><table><tbody><tr class=\\"bg-gray-100\\">
      <td class=\\"lg:px-4 border px-1 py-2 text-right text-xs\\">üèÜ</td>
      <td class=\\"lg:px-4 border px-1 py-2 text-right\\"><a rel=\\"noopener noreferrer\\" target=\\"_blank\\" href=\\"url\\">christianhaller</a></td>
      <td class=\\"lg:px-4 border px-1 py-2 text-right\\">4</td>
      <td class=\\"lg:px-4 border px-1 py-2 text-right\\">2</td>
      <td class=\\"lg:px-4 border px-1 py-2 text-right text-xs\\">50 years ago</td></tr><tr class=\\"bg-white\\">
      <td class=\\"lg:px-4 border px-1 py-2 text-right text-xs\\">ü•à</td>
      <td class=\\"lg:px-4 border px-1 py-2 text-right\\"><a rel=\\"noopener noreferrer\\" target=\\"_blank\\" href=\\"url\\">anneka</a></td>
      <td class=\\"lg:px-4 border px-1 py-2 text-right\\">99</td>
      <td class=\\"lg:px-4 border px-1 py-2 text-right\\">2</td>
      <td class=\\"lg:px-4 border px-1 py-2 text-right text-xs\\">50 years ago</td></tr><tr class=\\"bg-gray-100\\">
      <td class=\\"lg:px-4 border px-1 py-2 text-right text-xs\\">ü•â</td>
      <td class=\\"lg:px-4 border px-1 py-2 text-right\\"><a rel=\\"noopener noreferrer\\" target=\\"_blank\\" href=\\"url\\">another user</a></td>
      <td class=\\"lg:px-4 border px-1 py-2 text-right\\">333</td>
      <td class=\\"lg:px-4 border px-1 py-2 text-right\\">2</td>
      <td class=\\"lg:px-4 border px-1 py-2 text-right text-xs\\">50 years ago</td></tr><tr class=\\"bg-white\\">
      <td class=\\"lg:px-4 border px-1 py-2 text-right text-xs\\">4.</td>
      <td class=\\"lg:px-4 border px-1 py-2 text-right\\"><a rel=\\"noopener noreferrer\\" target=\\"_blank\\" href=\\"url\\">user4</a></td>
      <td class=\\"lg:px-4 border px-1 py-2 text-right\\">100</td>
      <td class=\\"lg:px-4 border px-1 py-2 text-right\\">2</td>
      <td class=\\"lg:px-4 border px-1 py-2 text-right text-xs\\">50 years ago</td></tr></tbody></table></div></div>"
`;
